blpapi.ts

```diff
@@ -0,0 +1,728 @@
+        var that = this;
+        this.session.emit = (): boolean => {
+            return that.emit.apply(that, arguments);
+        };
+    }
+
+    // MANIPULATORS
+    start(cb?: (err: any, value: any) => void): Promise<void> {
+        this.validateSession();
+
+        return new Promise<void>((resolve: Function, reject: Function): void => {
+            trace('Starting session');
+            this.invoke(this.session.start);
+
+            var listener = (listenerName: string, handler: Function, ev: any): void => {
+                this.removeAllListeners(listenerName);
+                handler(ev.data);
+            };
+
+            this.once('SessionStarted',
+                      listener.bind(this, 'SessionStartupFailure', (data: any): void => {
+                          log('Session started');
+                          trace(data);
+                          resolve();
+                      }));
+
+            this.once('SessionStartupFailure',
+                      listener.bind(this, 'SessionStarted', (data: any): void => {
+                          log('Session start failure');
+                          trace(data);
+                          reject(new BlpApiError(data));
+                      }));
+        }).nodeify(cb);
+    }
+
+    stop(cb?: (err: any, value: any) => void): Promise<void> {
+        return this.stopped = this.stopped ||
+                              new Promise<void>((resolve: Function, reject: Function): void => {
+            log('Stopping session');
+            this.invoke(this.session.stop);
+            this.once('SessionTerminated', (ev: any): void => {
+                resolve();
+            });
+        }).nodeify(cb);
+    }
+
+
+    // 'request' function overloads.
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            label: string,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            label: string,
+            callback: IRequestCallback): void;
+
+    // 'request' function definition
+    request(uri: string,
+            requestName: string,
+            request: any,
+            arg3: Object | string | IRequestCallback,
+            arg4?: string | IRequestCallback,
+            arg5?: IRequestCallback): void
+    {
+        var args = Array.prototype.slice.call(arguments);
+        // Get callback and remove it from array.
+        var callback = args.splice(args.length - 1)[0];
+        // Get optional identity and label arguments (idx 3 and 4 if present).
+        var optionalArgs = args.slice(3);
```

Comment:
Personally, I find it clearer to enumerate the arguments explicitly when the new argument list is not a continuous subsequence of the current one.